### PR TITLE
Adds SQL export functionality

### PR DESCRIPTION
**Pull request by rotodd**
## Description
I am merging my SQL export feature branch to master. The feature branch:
- Implements the group of asynchronous methods that make up the entire export operation, which includes export job:
  - Creation (#782)
  - Acquisition (#812)
  - Updates, retrieval by ID and retrieval by hash (#846)
- Creates new files to manage the SQL export updates (#865):
  - 2.diff.sql, which holds all the new export logic and is run after 1.sql on schema initialization
  - 2.sql, a combination of 1.sql and 2.diff.sql, which is a standalone file that offers an alternate schema initialization path

## Demo
Please see #846.

## Related issues
Addresses [#AB72049](https://microsofthealth.visualstudio.com/Health/_workitems/edit/72049).

## Testing
- Extends export tests for Cosmos to cover SQL
- Adds new tests for the update export method and schema migration checks
- Tests the two different schema upgrade paths (#872)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.SqlServer/Resources.resx [hunk 1]
```diff
@@ -120,6 +120,9 @@
   <data name="InvalidContinuationToken" xml:space="preserve">
     <value>The provided continuation token is not valid.</value>
   </data>
+  <data name="NullRowVersion" xml:space="preserve">
+    <value>The row version should never be null.</value>
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.SqlServer/Resources.resx]: Is there more meaning we can give when we need to debug this? Maybe also more context like `The update batch operation failed as no row version was returned`

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaInitializer.cs [hunk 2]
```diff
@@ -48,22 +48,36 @@ private void Initialize()
...
+                // And if we are applying a complete snapshot of the SQL schema
+                if (applySqlSchemaSnapshot)
+                {
+                    // Apply the maximum supported version. This won't consider the .diff.sql files.
+                    _schemaUpgradeRunner.ApplySchema((int)_schemaInformation.MaximumSupportedVersion, isSnapshot: true);
+                }
+                else
+                {
+                    // Otherwise, run version 1. We'll use this as a base schema and apply .diff.sql files to upgrade the schema version.
+                    _schemaUpgradeRunner.ApplySchema(version: 1, isSnapshot: true);
```
[the remark this answers is not in the document]
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaInitializer.cs]: Good point. I set the default path to upgrade using `.diff.sql` files, considering the scenario where a previous schema version is already running. 

However, if we find that the schema hasn't yet been initialized, I see that it makes sense to follow the snapshot path. Looking into this now.

File: src/Microsoft.Health.Fhir.SqlServer/Resources.resx (hunk 1 above)
Comment (johnstairs) [reply src/Microsoft.Health.Fhir.SqlServer/Resources.resx]: It's something that shouldn't even be possible. So I'm not sure the error message even needs to be in a resource file.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirOperationDataStore.cs [hunk 3]
```diff
@@ -5,40 +5,226 @@
...
+                    }
+                }
+
+                return acquiredJobs;
+            }
+        }
+
+        private static ExportJobOutcome CreateExportJobOutcome(string rawJobRecord, byte[] rowVersionAsBytes)
+        {
+            var exportJobRecord = JsonConvert.DeserializeObject<ExportJobRecord>(rawJobRecord);
```
[the remark this answers is not in the document]
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirOperationDataStore.cs]: Thanks, added it in 61ed541.

Just a note: this is passed in during serialization to ensure that the `Status` enum serializes as a string. Functionality doesn't seem to change whether or not it is used in deserialization, but I can see how it looks cleaner and could potentially change how things are handled on the back end.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaInitializer.cs [hunk 4]
```diff
@@ -37,7 +37,7 @@ public SchemaInitializer(SqlServerDataStoreConfiguration sqlServerDataStoreConfi
             _logger = logger;
         }
 
-        private void Initialize()
+        public void Initialize(bool applySqlSchemaSnapshot = false)
```
[the remark this answers is not in the document]
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaInitializer.cs]: Makes sense, thanks. Fixed in 4dc8c17.

File: src/Microsoft.Health.Fhir.SqlServer/Resources.resx (hunk 1 above)
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Resources.resx]: Removed it as a resource and updated the error message in e937c26.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaInitializer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaInitializer.cs]: Fixed in bfa5fb6.

**Pull request merged**